feat(curriculum): add Sudoku solver project Python (#52452)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/20-upcoming-python/learn-classes-and-objects-by-building-a-sudoku-solver/6569e37ec28e853628f18a86.md

---
id: 6569e37ec28e853628f18a86
title: Step 57
challengeType: 20
dashedName: step-57
---

# --description--

> Add all the variables to an iterable and pass it in the `all` function.
> `all` would make sure that all of the validation results are True.
> 
> Recall that the syntax is as follows:
>     
> ```js
> all([variable1, variable2, variable3])
> ```

Reviewer: ```suggestion
Insert `valid_in_row`, `valid_in_col`, and `valid_in_square` into a list and pass it to the `all()` function. This will verify that all the function calls return `True`.
```

`all` has been already used several times.

# --hints--

You should pass the variables `valid_in_row`, `valid_in_col`, and `valid_in_square` to the `all` function.

```js

> assert.match(code, /all\(\s*\[\s*valid_in_row\s*,\s*valid_in_col\s*,\s*valid_in_square\s*\]\s*\)/);

Reviewer: ```suggestion
({ test: () => assert.match(e.code.original["main.py"], /^\s{8}all\s*\(\s*\[\s*(valid_in_)(row|col|square)\s*,\s*\1(?!\2)(row|col|square)\s*,\s*\1(?!\2|\3)(row|col|square)\s*\]\s*\)/m) })
```
Allowing different variables order in the list.

            board_string += row_empty

            if index < 8:
                if index % 3 == 2:
                    board_string += f'╠═══{"╪═══"*2}{"╬═══"}{"╪═══"*2}{"╬═══"}{"╪═══"*2}╣\n'
                else:
                    board_string += middle_lines
            else:
                board_string += lower_lines

        return board_string

    def find_empty_cell(self):
        for row, contents in enumerate(self.board):
            try:
                col = contents.index(0)
                return row, col
            except ValueError:
                pass
        return None

    def valid_in_row(self, row, num):
        return num not in self.board[row]

    def valid_in_col(self, col, num):
        return all(
            self.board[row][col] != num
            for row in range(9)
        )

    def valid_in_square(self, row, col, num):
        row_start = (row // 3) * 3
        col_start=(col // 3) * 3
        for row_no in range(row_start, row_start + 3):
            for col_no in range(col_start, col_start + 3):
                if self.board[row_no][col_no] == num:
                    return False
        return True

>     def is_valid(self, empty, num):
>         row, col = empty
>         valid_in_row = self.valid_in_row(row, num)
>         valid_in_col = self.valid_in_col(col, num)
>         valid_in_square = self.valid_in_square(row, col, num)
> 
> --fcc-editable-region--
> 
> --fcc-editable-region--

Reviewer: ```suggestion
    def is_valid(self, empty, num):
        row, col = empty
--fcc-editable-region--        
        valid_in_row = self.valid_in_row(row, num)
        valid_in_col = self.valid_in_col(col, num)
        valid_in_square = self.valid_in_square(row, col, num)
        
--fcc-editable-region--
```
